File: server/DroidOSS.Core/Protocol.cs
namespace DroidOSS.Core;

/// <summary>
/// The wire format shared by the server, the phone, and the test tools.
/// </summary>
/// <remarks>
/// These values are transcribed from <c>docs/PROTOCOL.md</c>, which is the
/// canonical specification. They are implemented by hand in three languages —
/// C# here, Kotlin in the Android app, Python in <c>tools/fake_phone.py</c> —
/// and a disagreement between them produces no error at all, just a controller
/// that behaves strangely.
///
/// Change anything here and you must change all three, and bump <see cref="Version"/>.
/// </remarks>
public static class Protocol
{
    /// <summary>
    /// First byte of every packet. Anything without it is not ours.
    /// </summary>
    /// <remarks>
    /// A UDP socket receives port scans, other applications' stray broadcasts,
    /// and traffic from someone who mistyped an address. Without a marker byte
    /// that garbage would be read as controller state.
    /// </remarks>
    public const byte MagicByte = 0xDA;

    /// <summary>
    /// Format version, so an old phone and a new server can refuse each other
    /// politely rather than misreading the layout.
    /// </summary>
    public const byte Version = 1;

    /// <summary>UDP port carrying input and session messages.</summary>
    public const int InputPort = 27500;

    /// <summary>
    /// UDP port for discovery broadcasts only, kept separate so broadcast
    /// traffic never reaches the input socket's hot path.
    /// </summary>
    public const int DiscoveryPort = 27501;

    /// <summary>Total size of an INPUT packet: 4 header + 4 sequence + 12 payload.</summary>
    public const int InputPacketSize = 20;

    /// <summary>Size of the header every message shares, whatever its type.</summary>
    public const int HeaderSize = 4;

    /// <summary>
    /// Size of HELLO, WELCOME and BYE. They carry no payload — the header says
    /// everything they need to.
    /// </summary>
    public const int SessionMessageSize = HeaderSize;

    /// <summary>
    /// The <c>pad</c> byte when it names no particular slot.
    /// </summary>
    /// <remarks>
    /// In HELLO it means "any slot, you choose" — a phone never picks its own,
    /// or two phones both claim pad 0. In WELCOME it means the opposite end of
    /// the same conversation: "no slot for you", all four are in use.
    /// </remarks>
    public const byte NoPad = 0xFF;

    /// <summary>
    /// How long a session survives without a packet before the pad is zeroed
    /// and unplugged.
    /// </summary>
    /// <remarks>
    /// The phone sends every 8 ms, so 2 s is 250 missed packets — unambiguous
    /// silence rather than a rough patch of Wi-Fi. This is why the Android app
    /// must keep a send-rate floor even when idle: throttle below one packet per
    /// two seconds and the server would disconnect a phone sitting in a menu.
    /// </remarks>
    public static readonly TimeSpan SessionTimeout = TimeSpan.FromSeconds(2);

    /// <summary>
    /// How long a client waits for WELCOME before sending HELLO again.
    /// </summary>
    /// <remarks>
    /// UDP does not guarantee delivery, so a HELLO can simply vanish and the
    /// server would never know a phone was there. The client retries — it is the
    /// side that wants something. This also covers the ordinary case of the app
    /// being opened before the server is running.
    /// </remarks>
    public static readonly TimeSpan HelloRetryInterval = TimeSpan.FromMilliseconds(200);

    /// <summary>Byte positions within a packet. The codec never contains a bare number.</summary>
    public static class Offset
    {
        public const int Magic = 0;
        public const int Version = 1;
        public const int Type = 2;
        public const int Pad = 3;
        public const int Sequence = 4;

        // The payload below is byte-for-byte XINPUT_GAMEPAD — see PadState.
        public const int Buttons = 8;
        public const int LeftTrigger = 10;
        public const int RightTrigger = 11;
        public const int ThumbLX = 12;
        public const int ThumbLY = 14;
        public const int ThumbRX = 16;
        public const int ThumbRY = 18;
    }
}

/// <summary>
/// What a packet is for. Byte 2 of every message.
/// </summary>
/// <remarks>
/// There is deliberately no heartbeat type. The input stream is the heartbeat —
/// a packet every 8 ms means silence is unmistakable, which is what the
/// disconnect timeout relies on.
/// </remarks>
public enum MessageType : byte
{
    /// <summary>The 20-byte state snapshot. Phone to PC, 125 times a second.</summary>
    Input = 0x01,

    /// <summary>"I'm here." Phone to PC. The server assigns a pad slot in reply.</summary>
    Hello = 0x02,

    /// <summary>"You're pad 2." PC to phone. Also proves a server exists at this address.</summary>
    Welcome = 0x03,

    /// <summary>Clean exit. Phone to PC, so the pad unplugs now rather than on timeout.</summary>
    Bye = 0x04,

    /// <summary>Vibration intensity from the game. PC to phone.</summary>
    Rumble = 0x05,

    /// <summary>"Any servers out there?" Broadcast, answered with <see cref="Welcome"/>.</summary>
    Discover = 0x06,
}

File: androidapp3/app/src/main/java/io/github/kakashi812/droidoss/protocol/Protocol.kt
package io.github.kakashi812.droidoss.protocol

/**
 * The wire format, transcribed from `docs/PROTOCOL.md`.
 *
 * This is the third hand-written implementation of the same specification —
 * C# in `DroidOSS.Core/Protocol.cs`, Python in `tools/fake_phone.py`, Kotlin
 * here. Nothing generates one from another, so a disagreement produces no
 * error at all: just a controller that behaves strangely.
 *
 * Change anything here and you must change all three, and bump [VERSION].
 */
object Protocol {

    /** First byte of every packet. Anything without it is not ours. */
    const val MAGIC_BYTE: Byte = 0xDA.toByte()

    /** Format version, so an old phone and a new server refuse each other politely. */
    const val VERSION: Byte = 1

    /** UDP port carrying input and session messages. */
    const val INPUT_PORT = 27500

    /** UDP port for discovery broadcasts only. Not used until B6. */
    const val DISCOVERY_PORT = 27501

    /** 4 header + 4 sequence + 12 payload. */
    const val INPUT_PACKET_SIZE = 20

    /** The header every message shares, whatever its type. */
    const val HEADER_SIZE = 4

    /** HELLO, WELCOME and BYE carry no payload — the header says everything. */
    const val SESSION_MESSAGE_SIZE = HEADER_SIZE

    /**
     * The `pad` byte when it names no particular slot.
     *
     * In HELLO it means "any slot, you choose" — a phone never picks its own, or
     * two phones both claim pad 0. In WELCOME it means "no slot for you": all
     * four are in use.
     */
    const val NO_PAD: Byte = 0xFF.toByte()

    /** How long the server tolerates silence before zeroing and unplugging the pad. */
    const val SESSION_TIMEOUT_MS = 2_000L

    /**
     * How long to wait for WELCOME before sending HELLO again.
     *
     * UDP does not guarantee delivery, so a HELLO can simply vanish and the
     * server would never know this phone exists. The client retries because the
     * client is the side that wants something. This also covers the ordinary
     * case of the app being opened before the server is running.
     */
    const val HELLO_RETRY_MS = 200L

    /**
     * Packets per second.
     *
     * Fixed-rate rather than send-on-change: event-driven sending arrives in
     * bursts, which feels like jitter, and it makes silence impossible to tell
     * apart from "nothing moved" — which the disconnect timeout depends on.
     */
    const val SEND_RATE_HZ = 125

    /** Byte positions within a packet. The codec never contains a bare number. */
    object Offset {
        const val MAGIC = 0
        const val VERSION = 1
        const val TYPE = 2
        const val PAD = 3
        const val SEQUENCE = 4

        // The payload below is byte-for-byte XINPUT_GAMEPAD, which is what lets
        // the server read straight off the wire into the driver with no
        // conversion at all.
        const val BUTTONS = 8
        const val LEFT_TRIGGER = 10
        const val RIGHT_TRIGGER = 11
        const val THUMB_LX = 12
        const val THUMB_LY = 14
        const val THUMB_RX = 16
        const val THUMB_RY = 18
    }
}

/**
 * What a packet is for. Byte 2 of every message.
 *
 * There is deliberately no heartbeat type. The input stream *is* the heartbeat —
 * a packet every 8 ms means silence is unmistakable.
 */
enum class MessageType(val id: Byte) {
    /** The 20-byte state snapshot. Phone to PC, 125 times a second. */
    INPUT(0x01),

    /** "I'm here." Phone to PC. The server assigns a pad slot in reply. */
    HELLO(0x02),

    /** "You're pad 2." PC to phone. Also proves a server exists at this address. */
    WELCOME(0x03),

    /** Clean exit. Phone to PC, so the pad unplugs now rather than on timeout. */
    BYE(0x04),

    /** Vibration intensity from the game. PC to phone. Not used until B8. */
    RUMBLE(0x05),

    /** "Any servers out there?" Broadcast, answered with WELCOME. Not used until B6. */
    DISCOVER(0x06);

    companion object {
        fun fromId(id: Byte): MessageType? = entries.firstOrNull { it.id == id }
    }
}

/**
 * Button bits, as they sit in bytes 8–9 of an INPUT packet.
 *
 * Bit 11 is unused. The D-pad occupies bits 0–3 and is **not** optional: it is
 * what platformers, 2D games and emulation are played with, and those are this
 * project's best use case.
 */
object GamepadButton {
    const val DPAD_UP = 0x0001
    const val DPAD_DOWN = 0x0002
    const val DPAD_LEFT = 0x0004
    const val DPAD_RIGHT = 0x0008
    const val START = 0x0010
    const val BACK = 0x0020
    const val LEFT_THUMB = 0x0040
    const val RIGHT_THUMB = 0x0080
    const val LEFT_SHOULDER = 0x0100
    const val RIGHT_SHOULDER = 0x0200
    const val GUIDE = 0x0400
    const val A = 0x1000
    const val B = 0x2000
    const val X = 0x4000
    const val Y = 0x8000
}
